packages/cosmic-swingset/src/launch-chain.js

```diff
@@ -547,6 +519,41 @@ export async function launch({
     if (setActivityhash) {
       setActivityhash(controller.getActivityhash());
     }
+  }
+
+  async function endBlock(blockHeight, blockTime, params, newActions) {
+    // This is called once per block, during the END_BLOCK event, and
+    // only when we know that cosmos is in sync (else we'd skip kernel
+    // execution). 'newActions' are the bridge/mailbox/etc events that
+    // cosmos stored up for delivery to swingset in this block.
+
+    // First, push all newActions onto the end of the inboundQueue,
+    // remembering that inboundQueue might still have work from the
+    // previous block
+    let actionNum = 0;
+    for (const action of newActions) {
+      const inboundNum = `${blockHeight}-${actionNum}`;
+      inboundQueue.push({ action, inboundNum });
+      actionNum += 1;
+      inboundQueueMetrics.incStat();
+    }
+
+    // We update the timer device at the start of each block, which might push
```

Comment:
```suggestion
    // We update the timer device before running the kernel, which might push
```

Reply:
Actually this very much is only done once per block. With https://github.com/Agoric/agoric-sdk/pull/6741 (which this refactor is extracted from), the kernel is ran twice, but the the timers only polled once.

Reply:
> only done once per block

What I was trying to point out is "at the start of each block" is not correct anymore.  Maybe "before running the kernel at the end of the block"?